src/openfl/display/BitmapData.hx

```diff
@@ -2105,36 +2105,63 @@ class BitmapData implements IBitmapDrawable {
 		
 	}
 	
-	
-	private function __renderCanvas (renderSession:RenderSession):Void {
+	#if (js && html5)
+	function __canBeDrawnToCanvas ():Bool {
 		
-		#if (js && html5)
-		if (!readable) return;
+		return image != null;
+		
+	} 
+	
+	
+	function __drawToCanvas (context:CanvasRenderingContext2D, transform:Matrix, roundPixels:Bool, pixelRatio:Float, scrollRect:Rectangle, useScrollRectCoords:Bool):Void {
 		
 		if (image.type == DATA) {
 			
 			ImageCanvasUtil.convertToCanvas (image);
 			
 		}
+
+		var scale = pixelRatio / this.__pixelRatio; // Bitmaps can have different pixelRatio than display, therefore we need to scale them properly
 		
-		var context = renderSession.context;
-		
-		if (__worldTransform == null) __worldTransform = new Matrix ();
-		
-		context.globalAlpha = 1;
-		var transform = __worldTransform;
+		if (roundPixels) {
+			
+			context.setTransform (transform.a * scale, transform.b, transform.c, transform.d * scale, Math.round (transform.tx * pixelRatio), Math.round  (transform.ty * pixelRatio));
+			
+		} else {
+			
+			context.setTransform (transform.a * scale, transform.b, transform.c, transform.d * scale, transform.tx * pixelRatio, transform.ty * pixelRatio);
+			
+		}
 		
-		if (renderSession.roundPixels) {
+		if (scrollRect == null) {
 			
-			context.setTransform (transform.a, transform.b, transform.c, transform.d, Math.round (transform.tx), Math.round (transform.ty));
+			context.drawImage (image.src, 0, 0);
 			
 		} else {
 			
-			context.setTransform (transform.a, transform.b, transform.c, transform.d, transform.tx, transform.ty);
+			var dx, dy;
+			if (useScrollRectCoords) {
+				dx = scrollRect.x;
+				dy = scrollRect.y;
+			} else {
+				dx = dy = 0;
+			}
```

Comment:
I don't like this `useScrollRectCoords` too much, and I think you agree ;) I also have our discussion in mind that we can maybe improve this by avoiding the offset applied to `__renderTransform`, but this is something we would need to check.
While we can also leave it like this, I thougth we could maybe also pass over `dx` and `dy` to the function, leave it 0 for Tilemap, and based on the `scrollRect` for Bitmap? What do you think?

Reply:
Not sure we can easily change `__renderTransform`, because it's used for global-to-local, hit-test and god knows what else, I'd rather not touch it for now. Passing `dx`/`dy` as arguments sounds nice, I'll change that.

Reply:
Meh, passing `dx` and `dy` from scroll rect as arguments means that we need to do a null-check for `bitmap.__scrollRect` at the call-site, is that really better?

Reply:
I see, maybe not really better. I'm also fine with leaving it like it is.

For `__renderTransform` I still think we should try to fix it at some point. Not now, but maybe we can try soon.